git pinning remote branches gives ambiguity error

I'm trying to pin a branch:

```
$ opam remove cpdf
$ opam pin cpdf git://github.com/avsm/cpdf-source#annotation-page-numbers
[ERROR] /Users/avsm/.opam/packages.dev/cpdf.1.7 exists although cpdf.1.7 is not installed. Removing it.
$ opam install cpdf --verbose
The following actions will be performed:
 - install cpdf.pinned
1 to install | 0 to reinstall | 0 to upgrade | 0 to downgrade | 0 to remove

=-=-= Installing cpdf.pinned =-=-=
cpdf.pinned Fetching git://github.com/avsm/cpdf-source#annotation-page-numbers
Initialized empty Git repository in /Users/avsm/.opam/4.01.0/packages.dev/cpdf.pinned/.git/
From git://github.com/avsm/cpdf-source
 * [new branch]      annotation-page-numbers -> origin/annotation-page-numbers
 * [new branch]      master     -> origin/master
From git://github.com/avsm/cpdf-source
 * [new tag]         v1.7       -> v1.7
fatal: ambiguous argument 'annotation-page-numbers': unknown revision or path not in the working tree.
Use '--' to separate paths from revisions
```

Notice the error at the end -- what's the right way to pin the remote branch?

--- a/src/repositories/opamGit.ml
+++ b/src/repositories/opamGit.ml
@@ -48,37 +48,42 @@ module Git = struct
             full
       )
 
+  let get_commits repo =
+    match snd repo.repo_address with
+    | None -> [ "refs/remotes/origin/master" ]
+    | Some c -> [ "refs/remotes/origin/"^c;
+                  "refs/tags/"^c;
+                  c ]
+
   let reset repo =
     let merge commit =
-      try OpamSystem.command [ "git" ; "reset" ; "--hard"; commit ]; true
+      try OpamSystem.command [ "git" ; "reset" ; "--hard"; commit; "--" ]; true
       with _ -> false in
-    let commit = match snd repo.repo_address with
-      | None   -> "master"
-      | Some c -> c in
     OpamFilename.in_dir repo.repo_root (fun () ->
-        if not (merge ("refs/remotes/origin/"^commit)) then
-          if not (merge ("refs/tags/"^commit)) then
-            if not (merge commit) then
-              OpamSystem.internal_error "Unknown revision: %s." commit
+        let ok =
+          List.fold_left (fun ok commit -> if ok then ok else merge commit)
+            false (get_commits repo) in
+        if not ok then OpamSystem.internal_error "Unknown revision: %s."
+            (match snd repo.repo_address with Some a -> a | None -> "master")
       )
 
   let diff repo =
     let diff commit =
       try Some (
-        OpamSystem.read_command_output ["git" ; "diff" ; commit ; "--name-only"])
+        OpamSystem.read_command_output ["git" ; "diff" ; "--name-only" ; commit ; "--" ])
       with _ -> None in
-    let commit = match snd repo.repo_address with
-      | None   -> "origin/master"
-      | Some c -> c in
     OpamFilename.in_dir repo.repo_root (fun () ->
-      match diff commit with
-      | Some [] -> false
-      | Some _  -> true
-      | None       ->
-        match diff ("origin/"^commit) with
+        let diff =
+          List.fold_left (fun r commit -> match r with
+              | None -> diff commit
+              | Some x -> Some x)
+            None (get_commits repo) in
+        match diff with
         | Some [] -> false
         | Some _  -> true
-        | None    -> OpamSystem.internal_error "Unknown revision: %s." commit
+        | None    ->
+          OpamSystem.internal_error "Unknown revision: %s."
+            (match snd repo.repo_address with Some a -> a | None -> "master")
     )
 
 end

